### PR TITLE
Add support for euclidean division and modulation

### LINKED ISSUE #159: Add support for euclidean division and modulation
Support for `div_euclid` and `rem_euclid` would be very useful, particularly for `vek`: https://github.com/yoanlcq/vek/issues/11.

### PR DESCRIPTION
**Pull request by SparrowLii**
Fixes #159 
Add support for euclidean division and modulation.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -48,7 +49,6 @@ pub use ops::wrapping::{
 };
 pub use pow::{checked_pow, pow, Pow};
 pub use sign::{abs, abs_sub, signum, Signed, Unsigned};
-
```
Comment (cuviper) [inline src/lib.rs:51]: nit: this removal is unrelated, and I'd prefer to have that line for separation.

File: src/ops/euclid.rs [hunk 2]
```diff
@@ -0,0 +1,327 @@
+use core::ops::{Div, Rem};
+pub trait DivEuclid: Sized + Div<Self, Output = Self> {
```
Comment (cuviper) [inline src/ops/euclid.rs:2]: Please add a blank line around the trait declarations.

File: src/ops/euclid.rs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    /// such that `self >= n * v`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use num_traits::DivEuclid;
+    ///
+    /// let a: i32 = 7;
+    /// let b: i32 = 4;
+    /// assert_eq!(DivEuclid::div_euclid(a,b), 1); // 7 > 4 * 1
```
Comment (cuviper) [inline src/ops/euclid.rs:17]: nit: this should be spaced `(a, b)` -- and there's a lot of similar non-standard formatting in places that `rustfmt` doesn't reach within docs and macros. Please go through and clean that up if you can.

File: src/ops/euclid.rs [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    /// let b: i32 = 4;
+    /// assert_eq!(RemEuclid::rem_euclid(a,b), 3);
+    /// assert_eq!(RemEuclid::rem_euclid(-a,b), 1);
+    /// assert_eq!(RemEuclid::rem_euclid(a,-b), 3);
+    /// assert_eq!(RemEuclid::rem_euclid(-a,-b), 1);
+    /// ```
+    fn rem_euclid(self, v: Self) -> Self;
+}
+macro_rules! div_euclid_int_impl {
+    ($trait_name:ident for $($t:ty)*) => {$(
```
Comment (cuviper) [inline src/ops/euclid.rs:51]: The parameterized `$trait_name` isn't helping us here. I think this should hardcode the trait name and combine `div`/`rem`, so it can be simply invoked `euclid_int_impl(i8 i16 i32 i64 isize)`.

File: src/ops/euclid.rs [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        impl $trait_name for $t {
+            #[inline]
+            fn rem_euclid(self, v: $t) -> Self {
+                self % v
+            }
+        }
+    )*}
+}
+div_euclid_int_impl!(DivEuclid for i8 i16 i32 i64);
+div_euclid_uint_impl!(DivEuclid for isize usize u8 u16 u32 u64);
```
Comment (cuviper) [inline src/ops/euclid.rs:104]: `isize` belongs in the signed group.

**Review by cuviper — CHANGES_REQUESTED** [review]
I think we can use `FloatCore` to avoid the `cfg` limitations. Otherwise, there are some style issues, but I think this looks pretty good!

**Comment by SparrowLii** [comment]
Thanks very much for reviewing. I have made the corresponding changes. 
As a Rust learner (even an open source learner)， I look forward to contributing to Rust.

File: src/ops/euclid.rs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    /// assert_eq!(Euclid::rem_euclid(a, -b), 3);
+    /// assert_eq!(Euclid::rem_euclid(-a, -b), 1);
+    /// ```
+    fn rem_euclid(self, v: Self) -> Self;
+}
+macro_rules! euclid_int_impl {
+    ($($t:ty)*) => {$(
+        impl Euclid for $t {
+            #[inline]
+            fn div_euclid(self, v: $t) -> Self {
```
Comment (zesterer) [inline src/ops/euclid.rs:54]: Any reason for this to not use the corresponding method on all integer types?

Comment (SparrowLii) [reply src/ops/euclid.rs:54]: That's what I did before.  But this will cause the test of version 1.8.0 to fail, as it shows that the parameter type of div_eculid/rem_eculid is incorrect (expecting &self, not self, or on the contrary). So I implemented a separate implementation to avoid these compatibility issues.

**Comment by XAMPPRocky** [comment]
@cuviper Is there anything left in this PR that needs to be changed? We'd really like this functionality for `rust-gpu` so that we can support `rem_euclid` on SPIR-V platforms. 🙂

**Review by cuviper — CHANGES_REQUESTED** [review]
I think the methods should take references, for the sake of `num-bigint` where it's a waste of allocations to consume values. Either that, or go fully generic with a type parameter RHS and and associated `type Output`, so they can be implemented for any mix of values and references. There are annoyance trade-offs both ways.

**Comment by SparrowLii** [comment]
I think it is reasonable to use reference types from the aspect of efficiency.  If a general type is used and the Output type is specified, we need to implement it separately for the several cases(self, &self, rhs, &rhs). This should probably be what a direct user-facing library should do. As a basic library, num-trait should maintain its concise style IMO

File: src/lib.rs [hunk 7]
```diff
@@ -33,13 +33,15 @@ pub use bounds::Bounded;
 #[cfg(any(feature = "std", feature = "libm"))]
 pub use float::Float;
 pub use float::FloatConst;
+pub use float::FloatCore;
```
[the remark this answers is not in the document]
Comment (SparrowLii) [reply src/lib.rs]: Thanks for explanation, has removed it.

File: src/ops/euclid.rs [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    #[test]
+    fn euclid_float() {
+        macro_rules! test_euclid {
+            ($($t:ident)+) => {
+                $(
+                    {
+                        let x: $t = 12.1;
+                        let y: $t = 3.2;
+                        assert!(Euclid::div_euclid(&x, &y) * y + Euclid::rem_euclid(&x, &y) - x
+                        <=46.4 * <$t as ::float::FloatCore>::epsilon());
```
Comment (cuviper) [inline src/ops/euclid.rs:264]:
```suggestion
                        <= 46.4 * <$t as ::float::FloatCore>::epsilon());
```

**Review by cuviper — COMMENTED** [review]
Thanks, it looks nice.

Maybe we should also add a combined `div_rem_euclid`? (and checked.) That would let you get both values with the divisions and conditions only checked once. All that probably gets optimized together with primitive values anyway, but it would matter for types like `BigInt`.
